### ISSUE TITLE
Execute query always contains DISTINCT

### ISSUE THREAD

**Issue by yuki2006** [ent/ent#999]
https://github.com/facebook/ent/blob/fb97ca66ac4b1cfe0b85f4f060c3996d55ca7c22/entc/gen/template/dialect/sql/query.tmpl#L103

Probably because of the specification of this line, there is no way to make it False.
(Although it is possible if I make custom template..)

Also, if the query contains DISTINCT, it may affect SQL performance, so I want to remove DISTINCT by default.

**Comment by a8m** [comment]
Hey @yuki2006 and thanks for raising up this issue.

> Also, if the query contains DISTINCT, it may affect SQL performance, so I want to remove DISTINCT by default.

I get the use case, but 2 points about this - we need to come up with some numbers to understand the performance impact, and you should also we aware that it can return duplicate records in some cases (can't think about specific example atm). 

> Probably because of the specification of this line, there is no way to make it False.
(Although it is possible if I make custom template..)

Yes, and we can move this block to a smaller template to make it easier and override and maintain in the future. 
If you think about a different solution, please suggest. Otherwise, you're welcome to contribute this change:

```
{{ define "dialect/sql/query/spec" }}
{{ $builder := pascal $.Scope.Builder }}
{{ $receiver := receiver $builder }}

func ({{ $receiver }} *{{ $builder }}) querySpec() *sqlgraph.QuerySpec {
	_spec := &sqlgraph.QuerySpec{
		Node: &sqlgraph.NodeSpec{
			Table: {{ $.Package }}.Table,
			Columns: {{ $.Package }}.Columns,
			ID: &sqlgraph.FieldSpec{
				Type: field.{{ $.ID.Type.ConstName }},
				Column: {{ $.Package }}.{{ $.ID.Constant }},
			},
		},
		From: {{ $receiver }}.sql,
		Unique: true,
	}
	if ps := {{ $receiver }}.predicates; len(ps) > 0 {
		_spec.Predicate = func(selector *sql.Selector) {
			for i := range ps {
				ps[i](selector)
			}
		}
	}
	if limit := {{ $receiver }}.limit; limit != nil {
		_spec.Limit = *limit
	}
	if offset := {{ $receiver }}.offset; offset != nil {
		_spec.Offset = *offset
	}
	if ps := {{ $receiver }}.order; len(ps) > 0 {
		_spec.Order = func(selector *sql.Selector) {
			for i := range ps {
				ps[i](selector, {{ $.Package }}.ValidColumn)
			}
		}
	}
	return _spec
}
{{ end }}
```

**Comment by shuogawa** [comment]
@a8m 
Thank you for the wonderful ORM.
I and @yuki2006 work in the same team.
Our team has customized the template.

![image](https://user-images.githubusercontent.com/660998/102030481-4e4f4880-3df6-11eb-802d-032529a4ade7.png)

**Comment by BrentChesny** [comment]
Also ran into a case where this vastly impacted query performance (~50ms vs multiple seconds). Currently using the same workaround as @shuogawa.

**Comment by zzwx** [comment]
Someone asked for the numbers - please read **However** portion too.

The most simple query for `sources, err := book.QuerySources().All(ctx)`, takes 106ms on Postgres 11, table containing 1 million records. That's **9.4 queries per second**.
```
2021-01-27 08:56:55.193 EST [33248] LOG:  duration: 0.043 ms  bind lrupsc_1_14: SELECT DISTINCT "sources"."id", "sources"."url", "sources"."mime_type", "sources"."node_id", "sources"."modified" FROM "sources" WHERE "book_sources" = $1
2021-01-27 08:56:55.193 EST [33248] DETAIL:  parameters: $1 = '1649'
2021-01-27 08:56:55.300 EST [33248] LOG:  duration: 106.534 ms  execute lrupsc_1_14: SELECT DISTINCT "sources"."id", "sources"."url", "sources"."mime_type", "sources"."node_id", "sources"."modified" FROM "sources" WHERE "book_sources" = $1
2021-01-27 08:56:55.300 EST [33248] DETAIL:  parameters: $1 = '1649'
```

**However** I noticed that even regular SELECT in this case wasn't robust either.

So I managed to speed it up by simply adding an index on the edge like this: `index.Edges("book")`. That tricked the database to perform `(0.024+0.030)ms` per query, which is about **18.500 queries per second**.

**Comment by rubensayshi** [comment]
tbh the internet is littered with issues with bad performance with DISTINCT, I really don't think it should be the default.

I think my https://github.com/facebook/ent/pull/1142 PR could be used, but with Unique defaulting to false and the PR slightly adjusted to enabling 2 easy ways to set Unique (either on `...Query` or with annotation)

**Comment by zzwx** [comment]
@rubensayshi, @a8m
My original assumption was also that `SELECT DISTINCT` is always bad. However, there's also a bunch of reports that `SELECT DISTINCT` performs the same as a regular `SELECT ALL` because the optimizer will do away with the step necessary for eliminating duplicates. My particular use case showed that, of course it would depend on the quality of the optimizer, so this will be certainly different from case to case.

To be on the safe side, however, at this stage of development, I would rather prefer all join queries to be `SELECT DISTINCT` by default (provided the library is not yet wise enough to predict it's not necessary), because this way it allows to spot the slow queries without compromising the expectation of uniqueness, and then make adjustments to the query or/and schema accordingly. Having duplicates is much worse IMHO.

In other words, the flag to attempt to try non-distinct **should be explicit**, as this will hint the caller to be ready for, possibly, non-unique results. If I had a chance to vote, I would've voted for the safest, rather than assumingly more robust approach.

Of course, all above wouldn't be needed, had the library been able to eliminate `DISTINCT` based on logic, which is not a trivial task, I'm guessing.

**Comment by rubensayshi** [comment]
so distinct is only neccesary for JOINs, though your argument that for simple queries the optimizer should be able to understand that is probably true

**Comment by a8m** [comment]
Hey guys and thanks a lot for the feedback here.

@shuogawa, I just added the option in https://github.com/ent/ent/pull/1371 to disable the `DISTINCT` option using a flag. Briefly, as before, `DISTINCT` is enabled by default, and you can disable it using the `Unique` option on the query-builder. 

```go
client.Pet.Query().Unique(false).All(ctx)
```

We didn't change the default behavior because it's a breaking change, and in my opinion it's preferred in most cases.
However, in the future, there will be an option to disable it to all query-builders by default using an interceptor defined in the schema (or in a mixin shared between all schemas). For example:

```go
func (T) Interceptors() []ent.Interceptor {
	return []ent.Interceptor{
		func(next ent.Interceptor) ent.Interceptor {
			return ent.InterceptFunc(func(ctx context.Context, q ent.Query) (ent.Value, error) {
				q.Unique(false)
				return next.Intercept(ctx, q)
			})
		},
	}
}
```

**Comment by rubensayshi** [comment]
@a8m any point in rebaseing my https://github.com/ent/ent/pull/1142 to set it through annotations or you want to hold out until you have the interceptors going?

**Comment by shuogawa** [comment]
@a8m 

I'm sorry to be late. It is working. Thank you very much 🙇

**Comment by zzwx** [comment]
Can this be closed then?

**Comment by rubensayshi** [comment]
@a8m I still think my PR to set it on by-default through an annotation would be a good and simple addition and avoids having to mess with templates: https://github.com/ent/ent/pull/1142

**Comment by masseelch** [comment]
The issue described here seems to be fixed. As for the linked PR I will have a look at it and dicuss :-)

**Comment by sullerandras** [comment]
```
	entClient.Intercept(
		intercept.Func(func(ctx context.Context, q intercept.Query) error {
			q.Unique(false)
			return nil
		}),
	)
```

If i add this interceptor then the query will not have the `DISTINCT` even if i call `.Unique(true)` in my query. I thought the interceptor would allow us to set the default value to be `false` but it seems like it overwrites the value that i set in my query.

Is there any way to achieve not having `distinct` added to the queries by default, only when i explicitly call `.Unique(true)`?

**Comment by a8m** [comment]
Thanks for the feedback, @sullerandras. https://github.com/ent/ent/pull/3238 should address this issue. One it's merged, you can use this as follows:

```go
client.User.Intercept(
	intercept.Func(func(ctx context.Context, q intercept.Query) error {
		// Skip setting Unique if the modifier was set explicitly.
		if entgo.QueryFromContext(ctx).Unique == nil {
			q.Unique(false)
		}
		return nil
	}),
)
```
